File: extractors/go/thirdparty.go
// thirdparty.go — [mod]/... node creation + allowed gating.
//
// Mirrors the JS side's [npm]/... convention and python's [pip]/... registry:
// one package node per import, per-name member nodes for named access, all
// deduped in a single registry. `allowed` gates inclusion — empty set → ZERO
// third-party nodes (the engine controls this via options.includeThirdPartyLibs).

package main

import "strings"

// std packages never contain a dot in the first path element (documented
// go.dev module-path convention: external module paths always do). This is
// the same rule `go list std` encodes.
func isStdImport(path string) bool {
	first := path
	if i := strings.Index(path, "/"); i >= 0 {
		first = path[:i]
	}
	return !strings.Contains(first, ".")
}

// runtime/devtool categories (mirrors python's RUNTIME_PACKAGES/DEVTOOL_PACKAGES —
// frontend uses this for the library panel).
var runtimePkgs = []string{
	"github.com/gin-gonic/gin", "github.com/labstack/echo", "github.com/gofiber/fiber",
	"gorm.io/gorm", "github.com/jackc/pgx", "github.com/lib/pq",
	"github.com/go-sql-driver/mysql", "go.mongodb.org/mongo-driver",
	"modernc.org/sqlite", "github.com/mattn/go-sqlite3", "github.com/redis/go-redis",
	"github.com/golang-jwt/jwt", "github.com/dgrijalva/jwt-go", "golang.org/x/crypto",
	"github.com/google/uuid", "github.com/gorilla/mux", "github.com/gorilla/handlers",
	"github.com/gorilla/sessions", "github.com/stretchr/testify/assert",
}

var devtoolPkgs = []string{
	"github.com/stretchr/testify/mock", "github.com/stretchr/testify/require",
	"github.com/golang/mock", "github.com/go-cmp/cmp", "github.com/google/go-cmp",
	"github.com/onsi/ginkgo", "github.com/onsi/gomega", "go.uber.org/mock",
	"github.com/vektra/mockery", "github.com/boumenot/gocover-cobertura",
}

type ThirdPartyRegistry struct {
	allowed map[string]bool
	rawDeps map[string]string
	lookup  *LookupMaps
}

func newThirdPartyRegistry(allowed []string, fp *Fingerprint, l *LookupMaps) *ThirdPartyRegistry {
	set := map[string]bool{}
	for _, a := range allowed {
		set[a] = true
	}
	return &ThirdPartyRegistry{allowed: set, rawDeps: fp.RawDependencies, lookup: l}
}

func (t *ThirdPartyRegistry) permitted(pkg string) bool { return t.allowed[pkg] }

func (t *ThirdPartyRegistry) baseMetadata(pkg string) map[string]any {
	version, ok := t.rawDeps[pkg]
	if !ok {
		version = "unknown"
	}
	category := "unknown"
	for _, r := range runtimePkgs {
		if pkg == r || strings.HasPrefix(pkg, r+"/") {
			category = "runtime"
			break
		}
	}
	if category == "unknown" {
		for _, d := range devtoolPkgs {
			if pkg == d || strings.HasPrefix(pkg, d+"/") {
				category = "devtool"
				break
			}
		}
	}
	return map[string]any{"isThirdParty": true, "packageVersion": version, "category": category}
}

// packageNode — [mod]/pkg node, or nil when not permitted. Cached per id.
func (t *ThirdPartyRegistry) packageNode(pkg string) map[string]any {
	if !t.permitted(pkg) {
		return nil
	}
	id := thirdPartyID(pkg)
	if n, ok := t.lookup.thirdPartyByID[id]; ok {
		return n
	}
	n := map[string]any{
		"id":        id,
		"name":      pkg,
		"type":      "THIRD_PARTY",
		"filePath":  id,
		"startLine": 0,
		"endLine":   0,
		"metadata":  t.baseMetadata(pkg),
	}
	t.lookup.thirdPartyByID[id] = n
	t.lookup.thirdPartyNodes = append(t.lookup.thirdPartyNodes, n)
	return n
}

// memberNode — [mod]/pkg::member (lazy member access), or nil when not
// permitted. The package node always exists first (parent).
func (t *ThirdPartyRegistry) memberNode(pkg, name string) map[string]any {
	if !t.permitted(pkg) {
		return nil
	}
	pkgNode := t.packageNode(pkg)
	if pkgNode == nil {
		return nil
	}
	id := thirdPartyID(pkg) + "::" + name
	if n, ok := t.lookup.thirdPartyByID[id]; ok {
		return n
	}
	pkgMeta := pkgNode["metadata"].(map[string]any)
	meta := map[string]any{
		"isThirdParty":    true,
		"packageVersion":  pkgMeta["packageVersion"],
		"category":        pkgMeta["category"],
		"parentPackageId": pkgNode["id"],
		"methodName":      name,
	}
	n := map[string]any{
		"id":        id,
		"name":      pkg + "." + name,
		"type":      "THIRD_PARTY",
		"filePath":  thirdPartyID(pkg),
		"startLine": 0,
		"endLine":   0,
		"metadata":  meta,
	}
	t.lookup.thirdPartyByID[id] = n
	t.lookup.thirdPartyNodes = append(t.lookup.thirdPartyNodes, n)
	return n
}

File: extractors/rust/src/thirdparty.rs
// thirdparty.rs — [crate]/name third-party registry + gating (mirrors
// go/thirdparty.go and python/third_party.py).
//
// Gating rule (non-negotiable contract): options.includeThirdPartyLibs (both
// spellings) is THE gate — absent/empty → ZERO third-party nodes. All
// detectors (imports/calls/inheritance) share ONE registry.

use crate::contract::{base_metadata, code_node, CodeNode, NODE_THIRD_PARTY};
use crate::fingerprint::ParsedManifest;
use std::collections::{HashMap, HashSet};

/// std crates — first path segment of a use path in this set → skip tier
/// (no node, no edge).
pub const STD_CRATES: [&str; 3] = ["std", "core", "alloc"];

pub struct ThirdPartyRegistry {
    /// gated (allowed) crate names — empty = gate everything
    allowed: HashSet<String>,
    /// crate name → node (id [crate]/name)
    crate_nodes: HashMap<String, CodeNode>,
    /// crate name::member → node (id [crate]/name::member)
    member_nodes: HashMap<String, CodeNode>,
}

impl ThirdPartyRegistry {
    pub fn new(allowed: &[String]) -> Self {
        ThirdPartyRegistry {
            allowed: allowed.iter().cloned().collect(),
            crate_nodes: HashMap::new(),
            member_nodes: HashMap::new(),
        }
    }

    /// Strict membership gate (mirrors go's `permitted`): absent/empty
    /// `allowed` → ZERO third-party nodes (the engine controls inclusion via
    /// options.includeThirdPartyLibs). Hyphen/underscore normalization for
    /// Cargo crate names (dep `actix-web` ↔ crate `actix_web`).
    pub fn gated(&self, crate_name: &str) -> bool {
        if self.allowed.contains(crate_name) {
            return true;
        }
        let underscored = crate_name.replace('-', "_");
        self.allowed
            .iter()
            .any(|a| a.replace('-', "_") == underscored)
    }

    /// crate node [crate]/name — None when not gated.
    pub fn crate_node(&mut self, crate_name: &str) -> Option<CodeNode> {
        if !self.gated(crate_name) {
            return None;
        }
        if let Some(n) = self.crate_nodes.get(crate_name) {
            return Some(n.clone());
        }
        let mut meta = base_metadata();
        meta.insert("crateName".to_string(), crate_name.into());
        let n = code_node(
            &crate::lookup::third_party_id(crate_name),
            "",
            crate_name,
            NODE_THIRD_PARTY,
            0,
            0,
            "",
            meta,
        );
        self.crate_nodes.insert(crate_name.to_string(), n.clone());
        Some(n)
    }

    /// member node [crate]/name::member (e.g. serde::Serialize) — None when
    /// not gated.
    pub fn member_node(&mut self, crate_name: &str, member: &str) -> Option<CodeNode> {
        if !self.gated(crate_name) {
            return None;
        }
        let key = format!("{}::{}", crate_name, member);
        if let Some(n) = self.member_nodes.get(&key) {
            return Some(n.clone());
        }
        let mut meta = base_metadata();
        meta.insert("crateName".to_string(), crate_name.into());
        meta.insert("member".to_string(), member.into());
        let id = format!("{}::{}", crate::lookup::third_party_id(crate_name), member);
        let n = code_node(&id, "", &key, NODE_THIRD_PARTY, 0, 0, "", meta);
        self.member_nodes.insert(key, n.clone());
        Some(n)
    }

    /// All created nodes (collected last in the pipeline — like go).
    pub fn all_nodes(&self) -> Vec<CodeNode> {
        let mut out: Vec<CodeNode> = self.crate_nodes.values().cloned().collect();
        out.extend(self.member_nodes.values().cloned());
        out
    }
}

/// First segment of a path (e.g. "serde" from "serde::Serialize").
pub fn first_segment(path: &str) -> &str {
    path.split("::").next().unwrap_or(path)
}

/// True when the first segment is a std crate (skip tier).
pub fn is_std(first: &str) -> bool {
    STD_CRATES.contains(&first)
}

/// True when the first segment names an external crate (in the manifest's
/// dependency set). NOT std, NOT a local module — decided by callers using
/// the module map.
pub fn is_external_crate(first: &str, mf: &ParsedManifest) -> bool {
    mf.is_external_crate(first)
}
